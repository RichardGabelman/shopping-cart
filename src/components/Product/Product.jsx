import { useState } from "react";
import styles from "./Product.module.css";
import CustomInput from "./CustomInput/CustomInput.jsx";
import PropTypes from "prop-types";

function Product({ product, updateProductQuantity }) {
  const { id, title, image, price } = product;
  const [amount, setAmount] = useState(0);

  const handleAdd = () => {
    updateProductQuantity(id, amount);
    setAmount(0);
  };

  return (
    <article className={styles.productCard}>
      <img src={image} alt={title} className={styles.productImage} />
      <h2>{title}</h2>
      <p>${price.toFixed(2)}</p>
      <CustomInput amount={amount} setAmount={setAmount}></CustomInput>
      <button onClick={handleAdd}>Add to Cart</button>
    </article>
  );
}

Product.propTypes = {
  product: PropTypes.object.isRequired,
  updateProductQuantity: PropTypes.func.isRequired,
}

export default Product;
