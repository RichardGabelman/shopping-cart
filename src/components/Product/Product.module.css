.productCard {
  border: 1px solid black;
  display: flex;
  flex-direction: column;
  justify-content: end;
  align-items: center;
  width: 300px;
  height: 360px;
  padding: 5px;
  gap: 4px;
}

.productCard h2 {
  font-size: 20px;
  text-align: center;
}

.productCard p {
  font-size: 18px;
  min-width: 3ch;
  text-align: center;
}

.productCard > .productImage {
  width: 90px;
}