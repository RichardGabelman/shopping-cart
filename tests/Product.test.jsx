import { describe, it, expect, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Product from "../src/components/Product/Product";

describe("Product", () => {
  it("clicking add to cart calls updateProductQuantity", async () => {
    const user = userEvent.setup();

    const product = { id: 1, title: "First Item", image: "/img.png", price: 1.50};
    const updateProductQuantity = vi.fn();

    render(<Product product={product} updateProductQuantity={updateProductQuantity}/>);

    const addToCartBtn = screen.getByText("Add to Cart");

    await user.click(addToCartBtn);

    expect(updateProductQuantity).toHaveBeenCalledWith(1, 0);
  });
  it("clicking add to cart calls updateProductQuantity with correct amount", async () => {
    const user = userEvent.setup();

    const product = { id: 1, title: "First Item", image: "/img.png", price: 1.50};
    const updateProductQuantity = vi.fn();

    render(<Product product={product} updateProductQuantity={updateProductQuantity}/>);

    const plusBtn = screen.getByText("+");
    const addToCartBtn = screen.getByText("Add to Cart");
    const amountInput = screen.getByRole("spinbutton");

    expect(amountInput).toHaveValue(0);

    await user.click(plusBtn);

    expect(amountInput).toHaveValue(1);

    await user.click(addToCartBtn);

    expect(updateProductQuantity).toHaveBeenCalledWith(1, 1);

    expect(amountInput).toHaveValue(0);
  });
})