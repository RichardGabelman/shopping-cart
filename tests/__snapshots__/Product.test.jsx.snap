// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Product > renders correctly (snapshot) 1`] = `
<div>
  <article
    class="_productCard_3c78d5"
  >
    <img
      alt="First Item"
      class="_productImage_3c78d5"
      src="/img.png"
    />
    <h2>
      First Item
    </h2>
    <p>
      $
      1.50
    </p>
    <div
      class="_qtyContainer_e2dd4f"
    >
      <button>
        -
      </button>
      <input
        class="_input_e2dd4f"
        type="number"
        value="0"
      />
      <button>
        +
      </button>
    </div>
    <button>
      Add to Cart
    </button>
  </article>
</div>
`;
